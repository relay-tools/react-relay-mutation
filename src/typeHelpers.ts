export type Omit<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>>;

export type WithOptionalFields<T, K extends keyof T> = Omit<T, K> &
  Partial<Pick<T, K>>;
