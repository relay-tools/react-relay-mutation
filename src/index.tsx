import invariant from 'invariant';
import React, { useCallback, useContext, useState } from 'react';
import { ReactRelayContext, commitMutation } from 'react-relay';
import {
  MutationConfig as BaseMutationConfig,
  Environment,
  OperationBase,
} from 'relay-runtime';

import { Omit, WithOptionalFields } from './typeHelpers';

export type MutationState<T extends OperationBase> = {
  loading: boolean;
  data: T['response'] | null;
  error?: Error | null;
};

type MutationConfig<T extends OperationBase> = WithOptionalFields<
  Omit<BaseMutationConfig<T>, 'mutation'>,
  'variables'
>;

export type Mutate<T extends OperationBase> = (
  config?: Partial<MutationConfig<T>>,
) => Promise<T['response']>;

export function useMutation<T extends OperationBase>(
  mutation: BaseMutationConfig<T>['mutation'],
  userConfig: MutationConfig<T> = {},
  /** if not provided, the context environment will be used. */
  environment?: Environment,
): [Mutate<T>, MutationState<T>] {
  const [state, setState] = useState<MutationState<T>>({
    loading: false,
    data: null,
    error: null,
  });

  const relayContext = useContext(ReactRelayContext);
  const resolvedEnvironment = environment || relayContext.environment;
  const {
    configs,
    variables,
    uploadables,
    onCompleted,
    onError,
    optimisticUpdater,
    optimisticResponse,
    updater,
  } = userConfig;

  const mutate = useCallback(
    (config?: Partial<MutationConfig<T>>) => {
      const mergedConfig = {
        configs,
        variables,
        uploadables,
        onCompleted,
        onError,
        optimisticUpdater,
        optimisticResponse,
        updater,
        ...config,
      };

      invariant(mergedConfig.variables, 'you must specify variables');

      setState({
        loading: true,
        data: null,
        error: null,
      });

      return new Promise((resolve, reject) => {
        function handleError(error: any) {
          setState({
            loading: false,
            data: null,
            error,
          });

          if (!mergedConfig.onError) {
            reject(error);
          } else {
            mergedConfig.onError(error);
          }
        }
        commitMutation(resolvedEnvironment, {
          ...mergedConfig,
          mutation,
          variables: mergedConfig.variables!,
          onCompleted: (response, error) => {
            if (error) {
              handleError(error);
              return;
            }

            setState({
              loading: false,
              data: response,
              error: null,
            });

            resolve(response);
            if (mergedConfig.onCompleted) {
              mergedConfig.onCompleted(response, undefined);
            }
          },
          onError: handleError,
        });
      });
    },
    [
      resolvedEnvironment,
      configs,
      mutation,
      variables,
      uploadables,
      onCompleted,
      onError,
      optimisticUpdater,
      optimisticResponse,
      updater,
    ],
  );

  return [mutate, state];
}

export type MutationProps<T extends OperationBase> = MutationConfig<T> & {
  children: (mutate: Mutate<T>, state: MutationState<T>) => React.ReactNode;
  mutation: BaseMutationConfig<T>['mutation'];
  /** if not provided, the context environment will be used. */
  environment?: Environment;
};

export function Mutation<T extends OperationBase>({
  children,
  mutation,
  environment,
  ...config
}: MutationProps<T>) {
  const [mutate, state] = useMutation(mutation, config, environment);
  return children(mutate, state) as React.ReactElement<any>;
}
