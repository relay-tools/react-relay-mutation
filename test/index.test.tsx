import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import React from 'react';
import { graphql } from 'react-relay';
import { Environment } from 'relay-runtime';

import { Mutation, useMutation } from '../src';
import { indexTestMutation } from './__generated__/indexTestMutation.graphql';
import { createEnvironment, timeout } from './helpers';

Enzyme.configure({ adapter: new Adapter() });

let environment: Environment;

beforeEach(() => {
  environment = createEnvironment();
});

const mutation = graphql`
  mutation indexTestMutation($input: DoStuffInput!) {
    doStuff(input: $input) {
      result
    }
  }
`;

describe('render prop', () => {
  it('should fire onCompleted', async () => {
    const renderedStates: any[] = [];
    const onCompleted = jest.fn();
    const onError = jest.fn();

    const element = shallow(
      <Mutation<indexTestMutation>
        mutation={mutation}
        environment={environment}
        variables={{ input: { name: 'Mr. Foo' } }}
        onCompleted={onCompleted}
        onError={onError}
      >
        {(mutate, mutationState) => {
          renderedStates.push(mutationState);
          return (
            <button type="button" onClick={() => mutate()}>
              click
            </button>
          );
        }}
      </Mutation>,
    );

    element.find('button').simulate('click');

    await timeout(10);

    expect(onCompleted.mock.calls).toMatchSnapshot('onCompleted');
    expect(onError.mock.calls).toMatchSnapshot('onError');
    expect(renderedStates).toMatchSnapshot('state');
  });

  it('should fire onError', async () => {
    const onCompleted = jest.fn();
    const onError = jest.fn();
    const renderedStates: any[] = [];

    const element = shallow(
      <Mutation<indexTestMutation>
        mutation={mutation}
        environment={environment}
        variables={{ wrongInput: 1 } as any}
        onCompleted={onCompleted}
        onError={onError}
      >
        {(mutate, mutationState) => {
          renderedStates.push(mutationState);
          return (
            <button type="button" onClick={() => mutate()}>
              click
            </button>
          );
        }}
      </Mutation>,
    );

    element.find('button').simulate('click');

    await timeout(10);

    expect(onCompleted.mock.calls).toMatchSnapshot('onCompleted');
    expect(onError.mock.calls).toMatchSnapshot('onError');
    expect(renderedStates).toMatchSnapshot('state');
  });
});

describe('hook', () => {
  it('should work', async () => {
    const onCompleted = jest.fn();
    const onError = jest.fn();
    const renderedStates: any[] = [];

    const Component = () => {
      const [mutate, mutationState] = useMutation<indexTestMutation>(
        mutation,
        { onCompleted, onError },
        environment,
      );
      renderedStates.push(mutationState);

      return (
        <button
          type="button"
          onClick={() =>
            mutate({
              variables: { input: { name: 'Mr. Bar' } },
            })
          }
        >
          click me
        </button>
      );
    };

    const element = shallow(<Component />);

    element.find('button').simulate('click');

    await timeout(10);

    expect(onCompleted.mock.calls).toMatchSnapshot('onCompleted');
    expect(onError.mock.calls).toMatchSnapshot('onError');
    expect(renderedStates).toMatchSnapshot('state');
  });

  it('should succeed with promises', async () => {
    let called = false;

    const Component = () => {
      const [mutate] = useMutation<indexTestMutation>(
        mutation,
        {},
        environment,
      );

      return (
        <button
          type="button"
          onClick={async () => {
            const result = await mutate({
              variables: { input: { name: 'Mr. Bar' } },
            });
            expect(result).toMatchSnapshot();
            called = true;
          }}
        >
          click me
        </button>
      );
    };

    const element = shallow(<Component />);

    element.find('button').simulate('click');

    await timeout(10);

    expect(called).toBe(true);
  });

  it('should fail with promises', async () => {
    let called = false;

    const Component = () => {
      const [mutate] = useMutation<indexTestMutation>(
        mutation,
        {},
        environment,
      );

      return (
        <button
          type="button"
          onClick={async () => {
            try {
              await mutate({
                variables: { wrongInput: 1 } as any,
              });
            } catch (e) {
              expect(e).toMatchSnapshot();
              called = true;
            }
          }}
        >
          click me
        </button>
      );
    };

    const element = shallow(<Component />);

    element.find('button').simulate('click');

    await timeout(10);

    expect(called).toBe(true);
  });

  it('should resolve promise on error when onError is set', async () => {
    let called = false;

    const Component = () => {
      const [mutate] = useMutation<indexTestMutation>(
        mutation,
        {
          onError: jest.fn(),
        },
        environment,
      );

      return (
        <button
          type="button"
          onClick={async () => {
            const result = await mutate({
              variables: { wrongInput: 1 } as any,
            });
            expect(result).toMatchSnapshot();
            called = true;
          }}
        >
          click me
        </button>
      );
    };

    const element = shallow(<Component />);

    element.find('button').simulate('click');

    await timeout(10);

    expect(called).toBe(true);
  });
});
