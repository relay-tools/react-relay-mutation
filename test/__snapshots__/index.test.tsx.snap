// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`hook should fail with promises 1`] = `
[Error: Variable "$input" of non-null type "DoStuffInput!" must not be null.

GraphQL request:2:3
1 | mutation testIndexMutation(
2 |   $input: DoStuffInput!
  |   ^
3 | ) {]
`;

exports[`hook should resolve promise on error when onError is set 1`] = `undefined`;

exports[`hook should succeed with promises 1`] = `
Object {
  "doStuff": Object {
    "result": "you did it Mr. Bar",
  },
}
`;

exports[`hook should work: onCompleted 1`] = `
Array [
  Array [
    Object {
      "doStuff": Object {
        "result": "you did it Mr. Bar",
      },
    },
  ],
]
`;

exports[`hook should work: onError 1`] = `Array []`;

exports[`hook should work: state 1`] = `
Array [
  Object {
    "data": null,
    "error": null,
    "loading": false,
  },
  Object {
    "data": null,
    "error": null,
    "loading": true,
  },
  Object {
    "data": Object {
      "doStuff": Object {
        "result": "you did it Mr. Bar",
      },
    },
    "error": null,
    "loading": false,
  },
]
`;

exports[`render prop should fire onCompleted: onCompleted 1`] = `
Array [
  Array [
    Object {
      "doStuff": Object {
        "result": "you did it Mr. Foo",
      },
    },
  ],
]
`;

exports[`render prop should fire onCompleted: onError 1`] = `Array []`;

exports[`render prop should fire onCompleted: state 1`] = `
Array [
  Object {
    "data": null,
    "error": null,
    "loading": false,
  },
  Object {
    "data": null,
    "error": null,
    "loading": true,
  },
  Object {
    "data": Object {
      "doStuff": Object {
        "result": "you did it Mr. Foo",
      },
    },
    "error": null,
    "loading": false,
  },
]
`;

exports[`render prop should fire onError: onCompleted 1`] = `Array []`;

exports[`render prop should fire onError: onError 1`] = `
Array [
  Array [
    [Error: Variable "$input" of non-null type "DoStuffInput!" must not be null.

GraphQL request:2:3
1 | mutation testIndexMutation(
2 |   $input: DoStuffInput!
  |   ^
3 | ) {],
  ],
]
`;

exports[`render prop should fire onError: state 1`] = `
Array [
  Object {
    "data": null,
    "error": null,
    "loading": false,
  },
  Object {
    "data": null,
    "error": null,
    "loading": true,
  },
  Object {
    "data": null,
    "error": [Error: Variable "$input" of non-null type "DoStuffInput!" must not be null.

GraphQL request:2:3
1 | mutation testIndexMutation(
2 |   $input: DoStuffInput!
  |   ^
3 | ) {],
    "loading": false,
  },
]
`;
